BLD/DEV: ASAN at your fingertips!

lucascolley opened the pull request:
To simplify gh-24064.

This PR introduces a new `asan` env for `linux-64`, `linux-aarch64`, and `osx-arm64`, which pulls in `clangxx` and ASAN variants of NumPy and CPython, as defined in Pixi packages in their respective repos. The new Pixi tasks `build-asan` and `test-asan` can be used to develop with the address sanitizer with completely automated package builds and environment management.

Depends on https://github.com/numpy/numpy/pull/31805 -> https://github.com/python/cpython/pull/152385. Requires pixi >= v0.74.0.

Outstanding comments:
- workaround for linux crashes: https://github.com/scipy/scipy/pull/24066#discussion_r3720661996
- better way to pass `--capture=no` in `spin test`? https://github.com/scipy/scipy/pull/24066#discussion_r2566605635
- warning from Meson: https://github.com/scipy/scipy/pull/24066#discussion_r3523186617
	- it was suggested that we may just want to silence it: https://github.com/scipy/scipy/pull/24066#issuecomment-4882236497

Follow-up:
- add a sanitizers CI workflow: https://github.com/lucascolley/scipy/tree/asan-ci
- ccache for the Sanitizers workflow jobs

```console
scipy on 🎋 asan-stack [🔨] is 📦 v1.17.0.dev0 via 🐍 via 🧚 v0.60.0
❯ pixi run test-asan -t scipy/signal/tests/test_bsplines.py::TestSepfir2d::test_sepfir2d_strided_3
✨ Pixi task (build-asan in asan): spin build --asan --build-dir=build-asan --setup-args=-Dblas=blas --setup-args=-Dlapack=lapack --setup-args=-Duse-g77-abi=true: (Build SciPy, NumPy, and CPython with address sanitizer)
...
✨ Pixi task (test-asan in asan): spin test --no-build --build-dir=build-asan --no-capture -t scipy/signal/tests/test_bsplines.py::TestSepfir2d::test_sepfir2d_strided_3: (Test SciPy with address sanitizer)
...     
SUMMARY: AddressSanitizer: heap-buffer-overflow complex:781 in std::__1::complex<float> std::__1::operator*[abi:fe190107]<float, 0>(std::__1::complex<float> const&, std::__1::complex<float> const&)
Shadow bytes around the buggy address:
  0x611001f01f80: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
  0x611001f02000: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x611001f02080: 00 00 00 fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x611001f02100: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x611001f02180: fd fd fd fd fd fd fd fd fd fd fd fa fa fa fa fa
=>0x611001f02200: fa fa fa fa fa fa fa[fa]00 00 00 00 00 00 00 00
  0x611001f02280: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x611001f02300: 00 fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x611001f02380: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x611001f02400: 00 00 00 00 00 00 00 00 00 fa fa fa fa fa fa fa
  0x611001f02480: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==61562==ABORTING
Fatal Python error: Aborted
```

lucascolley commented on pixi.toml:
```diff
@@ -531,16 +535,33 @@ BOOST_INCLUDEDIR = "$PIXI_PROJECT_ROOT/.pixi/envs/py311-system-libs-osx/include/
 
 [feature.system-libs.tasks.test-system-libs]
-cmd = "spin test --no-build"
+cmd = "spin test --no-build --build-dir=build-system-libs"
```
fixes unrelated to this PR

lucascolley commented on .spin/cmds.py:
```diff
@@ -118,7 +118,14 @@ def build(*, parent_callback, meson_args, jobs, verbose, werror, asan, debug,
                                         "Please also check CXXFLAGS and FFLAGS.")
 
     if asan:
-        meson_args = meson_args + ('-Db_sanitize=address,undefined', )
```
as per https://github.com/scipy/scipy/pull/24064/files#r2566524626, using the builtin `-Db_sanitize` seems to try enabling sanitisation with `gfortran`, which errors out

FFY00 replied:
I feel like we should probably open a Meson bug about this, if we haven't already, because it shouldn't be trying to use the option if the compiler doesn't support it.

ngoldbaum replied:
Good call, I filed https://github.com/mesonbuild/meson/issues/15381

rgommers replied:
Can you add a code comment that this should be cleaned up as soon as we drop Fortran (2-3 months away by now) or https://github.com/mesonbuild/meson/issues/15381 gets fixed?

lucascolley commented:
@nickodell not sure if you have an osx-arm64 machine but you may be interested in checking this out (or extending it to linux)

lucascolley commented on pixi.toml:
```diff
@@ -531,16 +535,33 @@ BOOST_INCLUDEDIR = "$PIXI_PROJECT_ROOT/.pixi/envs/py311-system-libs-osx/include/
+description = "Build SciPy, NumPy, and CPython with address sanitizer"
+
+[feature.asan.tasks.test-asan]
+cmd = "spin test --no-build --build-dir=build-asan --no-capture"
```
```suggestion
cmd = "spin test --no-build --build-dir=build-asan --no-capture -m 'not fail_asan'"
```

lucascolley reviewed (commented):
as demonstrated in the latest commit, we can remove all of the pyenv, brew, pip, and GHA gfortran hacks malarkey from the CI workflow.

rgommers commented:
Looks promising! Is there an ETA for the `pixi-build` functionality becoming available in a `pixi` release?

My main concern is the git-stitching this into our main `pixi.toml`. WDYT about having `asan-stack` as a separate repo that in CI (and locally) can just be cloned and worked from? I think that may end up cleaner, and also that can then be used unchanged for CPython itself, for NumPy, for SciPy, for scikit-learn, etc.

ilayn commented:
> The issue with only compiling part of your stack with sanitizers is you will have both false positives and false negatives at library boundaries. IMO it's worth the internal consistency in the checks to always do a full-stack build to avoid sorting out issues that aren't really issues.

In the ideal world yes but there are many places in the stack where things are not sanitized. Especially if you are hunting for your own mistake among the cloud of output it becomes an impossible task to debug. I have been bitten by this in ARPACK and PROPACK translations multiple times which made me lost weeks staring at nonsensical output like the one at the top. You want things to fail immediately because otherwise you can't catch the first instance of the violation and then it spreads to other parts of the code. But if you fail immediately it never reaches your code and fails at some obscure Python.h line. 

Hence, I only add sanitization to that particular extension module to silence everything including other SciPy modules.

rgommers commented:
> But for this particular case (SciPy's in-tree workspace), I think it is really nice to also be able to make use of the git source — I think that contributors should be able to test with ASAN without needing to ever clone dependencies

My worry would be that we either end up pinning to some random old commit, or that we need to regularly bump the hash in `pixi.toml` which is work/churn. This is a quite advanced workflow, and I expect anyone using it to have no problem putting clones of `numpy` and `cpython` next to their `scipy` clone.

> In the ideal world yes but there are many places in the stack where things are not sanitized.

Not really, we only care about NumPy and CPython here (disclaimer: we don't do `openblas` here, and I suspect that's not clean), and both have ASan CI to keep things clean. And once we fix the remaining four suppressions, our own code base will also be clean. There's of course some chance of regressions, but enabling everything is generally the much more preferable strategy.

lucascolley commented:
> My worry would be that we either end up pinning to some random old commit, or that we need to regularly bump the hash in pixi.toml which is work/churn.

I think it makes sense to keep testing of unreleased versions of dependencies out-of-tree for now. But having `test-asan` point to the latest minor release tag of dependencies seems like not too much churn? That would be three bumps per year, or two if we do the autumn releases together?

rgommers commented:
> But having `test-asan` point to the latest minor release tag of dependencies seems like not too much churn?

Maybe. It may be more due to "find a pain point, fix it upstream, and then continue" work patterns. But we can always revisit, this isn't a hard to reverse decision. So let's give it a go your way I'd say.

lucascolley reviewed (commented):
@rgommers are you in principle happy with this approach now? If so, I guess we can push on with getting the upstream PRs merged, so that this isn't pointing to my fork by the time we merge to `main`?

rgommers commented on pixi.toml:
```diff
@@ -558,16 +562,42 @@ BOOST_INCLUDEDIR = "$PIXI_PROJECT_ROOT/.pixi/envs/py311-system-libs-osx/include/
+env.CC = "ccache $CC"
+env.CXX = "ccache $CXX"
+env.FC = "ccache $FC"
+env.ASAN_OPTIONS = "detect_leaks=0:symbolize=1:strict_init_order=true:allocator_may_return_null=1:use_sigaltstack=0"
```
`strict_init_order=true` is the only option that has an effect at build time, the rest can be dropped here.

FFY00 replied:
But we still need to make sure we can run Python code, right? From my experience, `detect_leaks=0` and `use_sigaltstack=0` may be needed just to run the Python interpreter, or import numpy (for `numpy-config`).

lucascolley replied:
> From my experience, `detect_leaks=0` and `use_sigaltstack=0` may be needed just to run the Python interpreter

added these back

FFY00 replied:
Just for reference, it's easy for `detect_leaks` to get false positives — on CPython, at least — or find real but unsuppressed issues. And for `use_sigaltstack`, both [`faulthandler`](https://docs.python.org/3/library/faulthandler.html) and the free-threading GC mess with the signal stack.

rgommers commented on pixi.toml:
```diff
@@ -558,16 +562,42 @@ BOOST_INCLUDEDIR = "$PIXI_PROJECT_ROOT/.pixi/envs/py311-system-libs-osx/include/
+[feature.asan.tasks.test-asan]
+cmd = "spin test --no-build --build-dir=build-asan --no-capture -m 'not fail_asan'"
+depends-on = "build-asan"
+env.ASAN_OPTIONS = "detect_leaks=0:symbolize=1:strict_init_order=true:allocator_may_return_null=1:use_sigaltstack=0"
```
`strict_init_order=true` can be removed here

rgommers replied:
I'm not 100% sure, since the docs are so terrible. Maybe just leave it in? I don't have time to investigate, and it's non-critical.

rgommers commented:
> @rgommers are you in principle happy with this approach now? If so, I guess we can push on with getting the upstream PRs merged, so that this isn't pointing to my fork by the time we merge to `main`?

Yes, I'm happy 👍🏼

lucascolley commented:
@FFY00 @rgommers do you have time to look at the outstanding comments here?

lucascolley committed:
BLD/DEV/CI: ASAN at your fingertips!

[skip circle]

andyfaff commented:
Why is the clang_asan job being removed? It's a job that checks that non-pixi spin workflows still work.

lucascolley commented:
> It's a job that checks that non-pixi spin workflows still work.

still work for what exactly? There are still other non-pixi jobs.

andyfaff commented:
> There are still other non-pixi jobs

Not on mac

lucascolley commented:
> > There are still other non-pixi jobs
> 
> Not on mac

fair enough! TBH I still don't understand the need for a non-pixi job for every platform. Are you concerned that e.g. there could be a bug in our code for the compilers on the image that we wouldn't hit on conda-forge compilers? Or is it moreso for the front-end, that e.g. the `spin` CLI still works?

andyfaff commented:
I think it's important to have workflows in CI that represent the workflow patterns of people that build locally. My workflow (on macOS) uses spin, doesn't use pixi or conda-forge packages. In addition, a major usage of scipy is via wheels, which are build in workflow that don't use conda packages, and are used in environments with non-conda packages. That clang_asan job keeps diversity.

lucascolley commented:
let's separate the discussion of workflow diversity out from this PR: https://github.com/scipy/scipy/issues/25015

lucascolley commented on pixi.toml:
```diff
@@ -782,6 +786,60 @@ depends-on = "build-system-libs"
+cxx-compiler = "*"
+
+[feature.asan.activation.env]
+FC = "ccache $CONDA_PREFIX/bin/gfortran"
```
```suggestion
FC = "ccache $FC"
```

lucascolley committed:
FC [skip circle]

lucascolley commented on scipy/signal/tests/test_bsplines.py:
```diff
@@ -338,7 +338,7 @@ def test_sepfir2d_strided_2(self, xp):
     @skip_xp_backends(np_only=True, reason="TODO: convert this test")
     @pytest.mark.xfail(reason="XXX: flaky. pointers OOB on some platforms")
-    @pytest.mark.fail_asan
+    # @pytest.mark.fail_asan
```
```suggestion
    @pytest.mark.fail_asan
```

lucascolley committed:
re-suppress test [skip circle]

lucascolley commented:
@nickodell looks like I got this working — ready for an analysis of timing, caching etc.

lucascolley commented on .github/workflows/asan.yml:
```diff
@@ -0,0 +1,50 @@
+        pixi-version: v0.65.0
+        cache: false
+        environments: asan
+        log-level: "q" # hide long CPython & NumPy build logs
```
```suggestion
        # log-level: "q" # hide long CPython & NumPy build logs
```

lucascolley commented:
I assume the CPython ASAN package is missing optimisations or is compiled with extra instrumentation compared to the previous pyenv build given how much slower the test-stage is — @FFY00 any idea?

lucascolley commented on pixi.toml:
```diff
@@ -782,6 +786,60 @@ depends-on = "build-system-libs"
+python.subdirectory = "Tools/pixi-packages/asan"
+numpy.git = "https://github.com/numpy/numpy.git"
+numpy.subdirectory = "pixi-packages/asan"
+gmpy2 = { version = "*", channel = "https://prefix.dev/sp-nightly-forge" }
```
it looks like the problem with Pythran is resolved, however I did have to build a Python `3.15.*` variant of `gmpy2` at https://github.com/lucascolley/sp-nightly-forge/blob/main/packages/gmpy2/pixi.toml to get this environment to resolve.

The problem isn't that we _need_ `gmpy2` in this env, but rather that if we were to remove it, we would have to drop the `test-deps` feature from the `asan` env, or split up the `test-deps` feature, e.g. into `noarch-test-deps` and `arch-specific-test-deps`, which doesn't feel great.

lucascolley commented:
> I assume the CPython ASAN package is missing optimisations or is compiled with extra instrumentation compared to the previous pyenv build given how much slower the test-stage is — @FFY00 any idea?

@crusaderky do you have experience of how the CPython Pixi packages compare to those one can acquire via `pyenv install` with `CONFIGURE_OPTS`?

ngoldbaum commented:
Total stab in the dark: are they built with `--enable-optimizations`? If not that will help performance at the cost of making the build slower. https://docs.python.org/3/using/configure.html#cmdoption-enable-optimizations

ngoldbaum commented:
That said, the cpython_sanity images don't enable any configure options besides the ones for sanitizers: https://github.com/nascheme/cpython_sanity/blob/main/Dockerfile.cpython

lucascolley commented:
> Total stab in the dark: are they built with `--enable-optimizations`? If not that will help performance at the cost of making the build slower. [docs.python.org/3/using/configure.html#cmdoption-enable-optimizations](https://docs.python.org/3/using/configure.html#cmdoption-enable-optimizations)

I don't think so https://github.com/python/cpython/blob/main/Tools/pixi-packages/build.sh

crusaderky commented:
> > I assume the CPython ASAN package is missing optimisations or is compiled with extra instrumentation compared to the previous pyenv build given how much slower the test-stage is — @FFY00 any idea?
> 
> @crusaderky do you have experience of how the CPython Pixi packages compare to those one can acquire via `pyenv install` with `CONFIGURE_OPTS`?

I am not aware of differences, but I have not worked with the pyenv ASAN build either so there may be some. If you find any differences I think it would be worth discussing them on the cpython issue tracker and potentially get rid of them.

tylerjereddy commented on .github/workflows/array_api.yml:
```diff
@@ -60,7 +60,7 @@ jobs:
       - uses: prefix-dev/setup-pixi@a0af7a228712d6121d37aba47adf55c1332c9c2e # v0.9.4
         with:
-          pixi-version: v0.60.0
+          pixi-version: v0.65.0
```
Here and elsewhere, you are bumping the `pixi` version on unrelated CI configs because you're using new `pixi` features in `pixi.toml` that require the bump? Or just unrelated bumps while in the neighborhood?

lucascolley replied:
The former. We should add https://pixi.prefix.dev/latest/reference/pixi_manifest/#requires-pixi-optional to make that clear.

tylerjereddy commented on .github/workflows/macos.yml:
```diff
@@ -90,74 +90,3 @@ jobs:
-      run: |
-        # pass -s to pytest to see ASAN errors and warnings, otherwise pytest captures them
-        ASAN_OPTIONS=detect_leaks=0:symbolize=1:strict_init_order=true:allocator_may_return_null=1:use_sigaltstack=0 \
-        spin test -- -v -s --timeout=600 --durations=10 -m 'not fail_asan'
```
did the team agree that this job was "ok" to remove? can you link to the discussion about that?

lucascolley replied:
Andrew did raise a concern but I believe it is addressed given https://github.com/scipy/scipy/issues/25015#issuecomment-4250643829.

tylerjereddy reviewed (commented):
Do you think this will be ready to merge in ~12 days before branching?

Is there a compelling reason to merge this very shortly before branching vs. after?

I added some minor comments--apologies if those were already discussed but the discussion is quite long here and there are many milestoned PRs.

lucascolley commented:
> Do you think this will be ready to merge in ~12 days before branching?
> 
> Is there a compelling reason to merge this very shortly before branching vs. after?

Perhaps not, and no, I don't think so — this is really just developer- rather than user-facing.

lucascolley commented:
something strange seems to be blocking us from using a recent version of clang on osx-arm64, https://github.com/prefix-dev/pixi/issues/6460

lucascolley commented on .spin/cmds.py:
```diff
@@ -123,7 +123,16 @@ def build(*, parent_callback, meson_args, jobs, verbose, werror, asan, debug,
+        root = Path(__file__).parent.parent
+        asan_ignore_file = root.absolute() / 'tools' / 'asan-ignore.txt'
+
+        # TODO: `-Db_sanitize=address,undefined` ?
+        meson_args += ('-Db_sanitize=address',)
```
also seeing the following warning from Meson

```
scipy/misc/meson.build:7: WARNING: Trying to use ['address'] sanitizer on Clang with b_lundef.
This will probably not work.
Try setting b_lundef to false instead.
```

lucascolley reviewed (commented):
this is now working for me locally on both linux-64 and osx-arm64, using `pixi` and `pixi-build-python` from the branch of https://github.com/prefix-dev/pixi/pull/6519. As soon as that is released this should be ready to review with feedback from CI.

Depends on:
- https://github.com/python/cpython/pull/152385
- https://github.com/numpy/numpy/pull/31805

ngoldbaum commented:
> also seeing the following warning from Meson
>
>scipy/misc/meson.build:7: WARNING: Trying to use ['address'] sanitizer on Clang with b_lundef.
This will probably not work.
Try setting b_lundef to false instead.

I don't know why meson does this. It's been generating this warning for NumPy ASAN builds for years.

I'm pretty sure it's spurious for our use-case here. It may be needed if you're not compiling the full application (e.g. CPython + an extension) with ASan, but we are.

If you'd like, feel free to open or look for an existing issue about this in Meson.

lucascolley commented:
also seeing various `-Wunused-result` from `_fast_matrix_market` when running `build-asan` on osx-arm64

lucascolley commented on .github/workflows/sanitizers.yml:
I guess we need to start using https://github.com/scipy/scipy/blob/main/.github/ccache/action.yml. Probably we also want https://github.com/scipy/scipy/blob/main/.github/workflows/linux.yml#L19-L20 and a `ccache -s` step.

lucascolley commented:
ping @rgommers !

rgommers commented on pixi.toml:
```diff
@@ -851,6 +856,58 @@ depends-on = "build-system-libs"
+clangxx_osx-arm64 = "*"
+
+[feature.asan.tasks.build-asan]
+# TODO: investigate why `use-pythran=false` is needed
```
Not blocking for merge, but this would be good to do - at least ensuring we update the comment to understand what's going wrong (building, specific error at runtime, ....?).

lucascolley replied:
> Not blocking for merge, but this would be good to do - at least ensuring we update the comment to understand what's going wrong (building, specific error at runtime, ....?).

with Pythran, `pixi run build-asan` fails with:

```
[76/1135] Generating 'scipy/linalg/_linalg_pythran.cpython-316-darwin.so.p/_linalg_pythran.cpp'
FAILED: [code=1] scipy/linalg/_linalg_pythran.cpython-316-darwin.so.p/_linalg_pythran.cpp
/Users/lucascolley/ghq/github.com/scipy/scipy/.pixi/envs/asan/bin/pythran -E ../scipy/linalg/_linalg_pythran.py -o scipy/linalg/_linalg_pythran.cpython-316-darwin.so.p/_linalg_pythran.cpp
...
 File "/Users/lucascolley/ghq/github.com/scipy/scipy/.pixi/envs/asan/lib/python3.16/site-packages/pythran/run.py", line 190, in run
    pythran.compile_pythranfile(args.input_file,
...
  File "/Users/lucascolley/ghq/github.com/scipy/scipy/.pixi/envs/asan/lib/python3.16/site-packages/pythran/optimizations/dead_code_elimination.py", line 124, in visit_If
    if ast.literal_eval(node.test):
       ~~~~~~~~~~~~~~~~^^^^^^^^^^^
  File "/Users/lucascolley/ghq/github.com/scipy/scipy/.pixi/envs/asan/lib/python3.16/site-packages/gast/gast.py", line 394, in literal_eval
    node_or_string = gast_to_ast(node_or_string)
  File "/Users/lucascolley/ghq/github.com/scipy/scipy/.pixi/envs/asan/lib/python3.16/site-packages/gast/ast3.py", line 514, in gast_to_ast
    return GAstToAst3().visit(node)
           ~~~~~~~~~~~~~~~~~~^^^^^^
  File "/Users/lucascolley/ghq/github.com/scipy/scipy/.pixi/envs/asan/lib/python3.16/ast.py", line 542, in visit
    return visitor(node)
  File "/Users/lucascolley/ghq/github.com/scipy/scipy/.pixi/envs/asan/lib/python3.16/site-packages/gast/astn.py", line 20, in generic_visit
    new_node = getattr(to, cls)()
TypeError: ast.Compare.__init__ missing 1 required positional argument: 'left'
```

rgommers replied:
@serge-sans-paille does this look familiar? Worth a bug report on either `pythran` or `gast`? Would be nice to get ASan to work with Pythran.

lucascolley replied:
may be a change in py316dev that requires adjustment?

serge-sans-paille replied:
yeah, please open a bug in gast with a 3.16 reproducer, that would be much appreciated.
This should be unrelated to ASAN

serge-sans-paille replied:
It doesn't mean pythran - generated code is ASAN proof though, but that's a topic I can start on my side.

lucascolley replied:
```suggestion
# `use-pythran=false` needed with py316dev due to https://github.com/serge-sans-paille/gast/issues/110
```

rgommers commented on .spin/cmds.py:
```diff
@@ -123,7 +123,16 @@ def build(*, parent_callback, meson_args, jobs, verbose, werror, asan, debug,
+        # TODO: `-Db_sanitize=address,undefined` ?
+        meson_args += ('-Db_sanitize=address',)
+        compiler_args = (
+            f'-fno-omit-frame-pointer -fsanitize-ignorelist={asan_ignore_file}'
```
I see this ignore is passed along correctly, but with `pixi r test -s signal` I immediate hit an error in `bsplines` (xref gh-24681) on Linux x86-64, which looks like the thing being ignored. The right move is just to fix that problem so we have zero ignores, then merge this PR, and keep it at zero.

rgommers replied:
This shows up in both CI jobs as well, they red with:
```
[762](https://github.com/scipy/scipy/actions/runs/28874377852/job/85645387625?pr=24066#step:5:41763)
scipy/signal/tests/test_bsplines.py::TestSepfir2d::test_sepfir2d_strided_3[numpy-float] XPASS
=================================================================
==45926==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6110096e95f8 at pc 0x000103d5a8b0 bp 0x00016f931830 sp 0x00016f931828
READ of size 4 at 0x6110096e95f8 thread T0
....
```

rgommers commented on .github/workflows/asan.yml:
```diff
@@ -0,0 +1,50 @@
+name: Address Sanitizer Tests
```
Can we use filename and title that's a bit more generic, just "Sanitizers"? That way we won't have to move the file and have this then-outdated name hanging around in the Actions logs/sidebar when we add the next sanitizer.

rgommers reviewed (commented):
This is very close, testing on Linux and it works as advertised. I also had a new crash report at hand, and it cleanly reproduced after upgrading pixi and running `pixi r test-asan`. 

CI is red; all the remaining comments are minor though, and `ccache` will be helpful but can wait till later indeed.

lucascolley commented on scipy/signal/tests/test_bsplines.py:
```diff
@@ -337,7 +337,7 @@ def test_sepfir2d_strided_2(self, xp):
     @skip_xp_backends(np_only=True, reason="TODO: convert this test")
     @pytest.mark.xfail(reason="XXX: flaky. pointers OOB on some platforms")
-    @pytest.mark.fail_asan
+    # @pytest.mark.fail_asan
```
thanks Ralf! This line here is why CI is (deliberately) red

rgommers replied:
Ah, that makes sense:) Guess that can be reverted then

rgommers commented:
@lucascolley can you update the PR description with the relevant info, more recent PRs and the Pixi version this depends on? Would be nice to get that description as self-contained and complete as possible; I suspect someone is going to revisit that to learn from it or for debugging.

lucascolley commented:
> @lucascolley can you update the PR description with the relevant info, more recent PRs and the Pixi version this depends on? Would be nice to get that description as self-contained and complete as possible; I suspect someone is going to revisit that to learn from it or for debugging.

Took a stab, please let me know what else would be helpful to include!

The remaining concern is how long these jobs will take to run, I suppose. Also, whether the `cache: true` I have included for setup-pixi at the minute is useful/workable.

rgommers commented:
Thanks, PR description looks good to me now. CI runtime and caches: not so much - I saw 22 minutes on the first run, but missed that that was aborting early. Both jobs have now been running for over an hour, which isn't workable. It's not using `pytest-xdist` though. Let's try optimizing that? Runners have 4 cores; ASan is a little memory-hungry, but `-j4` might still fit.

Let's worry about caches later.

lucascolley commented:
> Both jobs have now been running for over an hour, which isn't workable. It's not using `pytest-xdist` though. Let's try optimizing that? Runners have 4 cores; ASan is a little memory-hungry, but `-j4` might still fit.

`-j4` does not seem to have solved this problem.

rgommers commented:
The logs are too large, but from the raw logs you can see that the Linux job passed twice except for two failures that I believe are unrelated (I saw them in other jobs too recently, IIRC on i686, test is unstable):
```
FAILED scipy/sparse/linalg/_isolve/tests/test_iterative.py::test_convergence[rand-sym-pd-F-tfqmr-numpy-batch_b0-batch_A0] - assert 90 == 0
FAILED scipy/sparse/linalg/_isolve/tests/test_iterative.py::test_precond_dummy[rand-sym-pd-F-tfqmr-numpy-batch_b0-batch_A0] - assert 90 == 0
2026-07-10T16:05:28.3930579Z = 2 failed, 88114 passed, 8186 skipped, 314 xfailed, 20 xpassed in 3262.15s (0:54:22) =
```
The Linux job took 2h 13m in serial, and 1h 28m with `-j4`.

The macOS job passed in 1h 44m in serial, and timed out after 6 hours with `-j4`. The logs stop at this line, but I suspect that that's not meaningful and the problem is too much memory usage:
```
[gw1] PASSED scipy/special/tests/test_support_alternative_backends.py::test_support_alternative_backends[numpy-pdtri-intp-shapes1] 
##[error]The operation was canceled.
```

Maybe `-j2` would be best, but it's still going to take over an hour, which is too much. It'd be useful to add `--durations=50` to the test invocation to check what tests are really slow (it's likely that some are really slow, and it's not a uniform slowdown).

rgommers commented on pixi.toml:
```diff
@@ -851,6 +856,58 @@ depends-on = "build-system-libs"
 
 
+[feature.asan]
+platforms = ["osx-arm64", "linux-64"]
```
If it resolves on `linux-aarch64`, that would be useful to add here (it's my default platform more often these days, using containers on macOS arm64).

rgommers commented:
Once we know what the slowest tests are, we can skip them or tune `--timeout`. For `pytest-xdist` there's `--dist worksteal` which could help if one of the workers is way slower than the others (fairly likely).

Speeding up can also be done with `ASAN_OPTIONS` (in CI only, not for local builds). `malloc_context_size=7` looks most promising to me; the default is 30 frames, 7 should reduce memory usage a little bit and improve runtime performance. The impact is that diagnostics may sometimes miss the relevant frame when hitting an error, but it won't affect detection, only the traceback we get after.

On macOS, the hang could be avoided by reducing memory usage, with `quarantine_size_mb=64`.

lucascolley commented:
- switched to `-j2`
- added `quarantine_size_mb=64` for `osx-arm64`
- added `malloc_context_size=7` in CI
- bumped `--durations=10` to `--durations=50`

`linux-aarch64` is annoyingly blocked on `gmpy2` — I can try publishing a compatible binary to `sp-nightly-forge` soon.

lucascolley commented:
linux-64 run finished in 91mins, with the same flaky failures. Here are the slow tests:

<details>

<html><head></head><body><h1>Slowest 50 Test Durations</h1>

Duration | Type | Module | Test
-- | -- | -- | --
47.46s | call | scipy/_lib/tests/test_import_cycles.py | test_public_modules_importable_2
43.57s | setup | scipy/_lib/tests/test_warnings.py | test_warning_calls_filters
41.81s | call | scipy/special/tests/test_cdflib.py | TestCDFlib::test_chndtrix
41.56s | call | scipy/linalg/tests/test_extending.py | test_cython
38.86s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_e1_complex
31.93s | call | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[_LogUniform]
31.82s | call | scipy/stats/tests/test_mgc.py | TestMGCStat::test_pvalue_literature
30.38s | call | scipy/_lib/tests/test__util.py | TestTransitionToRNG::test_rng_deterministic[basinhopping-seed]
22.63s | call | scipy/stats/tests/test_multivariate.py | TestNormalInverseGamma::test_moments
21.58s | call | scipy/optimize/tests/test_extending.py | test_cython
21.32s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_bessely_complex
21.27s | call | scipy/_lib/tests/test__util.py | TestTransitionToRNG::test_rng_deterministic[differential_evolution-seed]
20.71s | call | scipy/optimize/tests/test__differential_evolution.py | TestDifferentialEvolutionSolver::test_L1
20.46s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_erfc_complex
20.31s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_legendre_int
19.60s | call | scipy/stats/tests/test_multivariate.py | TestMultivariateNormal::test_cdf_vs_cubature[3]
19.02s | call | scipy/stats/tests/test_stats.py | TestPageTrendTest::test_accuracy2[542-0.9481266260876332-True-exact-data0]
18.63s | call | scipy/special/tests/test_extending.py | test_cython
18.59s | call | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[StandardNormal]
18.16s | call | scipy/special/tests/test_cdflib.py | TestCDFlib::test_bdtrin
18.00s | call | scipy/optimize/tests/test_optimize.py | TestOptimizeSimple::test_minimize_callback_copies_array[fmin]
17.60s | call | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[Logistic]
16.72s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_erf_complex
16.68s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_pcfw
15.61s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_erfi_complex
14.98s | call | scipy/optimize/tests/test__basinhopping.py | TestBasinHopping::test_all_nograd_minimizers[COBYLA]
14.92s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_rgamma
14.29s | call | scipy/special/tests/test_cython_special.py | test_cython_api[elliprj]
14.20s | call | scipy/special/tests/test_mpmath.py | test_spence_circle
13.88s | call | scipy/optimize/tests/test_constraint_conversion.py | TestNewToOld::test_individual_constraint_objects
13.14s | call | scipy/special/tests/test_mpmath.py | test_rgamma_zeros
13.07s | call | scipy/interpolate/tests/test_fitpack2.py | TestRectBivariateSpline::test_spline_large_2d[spl_apis1-s_tols1-shape1]
12.90s | call | scipy/interpolate/tests/test_rbfinterp.py | TestRBFInterpolatorNeighbors20::test_concurrency
12.56s | call | scipy/special/tests/test_mpmath.py | test_erf_complex
12.52s | call | scipy/_lib/tests/test__util.py | TestTransitionToRNG::test_rng_deterministic[goodness_of_fit-random_state]
12.52s | call | scipy/special/tests/test_mpmath.py | test_beta
12.41s | call | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[Normal]
12.25s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_log_ndtr_complex
12.14s | call | scipy/interpolate/tests/test_fitpack2.py | TestRectBivariateSpline::test_spline_large_2d[spl_apis1-s_tols2-shape1]
11.99s | call | scipy/special/tests/test_cdflib.py | TestCDFlib::test_gdtrib
11.25s | call | scipy/special/tests/test_hyp2f1.py | TestHyp2f1::test_test_hyp2f1
11.24s | call | scipy/spatial/tests/test_distance.py | TestPdist::test_pdist_correlation_iris_nonC
11.09s | call | scipy/signal/tests/test_signaltools.py | test_filtfilt_gust
11.05s | call | scipy/optimize/tests/test_constraint_conversion.py | TestNewToOldCobyla::test_list_of_problems
10.52s | call | scipy/spatial/tests/test_distance.py | TestPdist::test_pdist_correlation_iris
10.45s | call | scipy/special/tests/test_cdflib.py | TestCDFlib::test_btdtria
10.40s | call | scipy/special/tests/test_mpmath.py | TestSystematic::test_erfi
10.30s | call | scipy/optimize/tests/test__differential_evolution.py | TestDifferentialEvolutionSolver::test_strategy_fn
9.86s | call | scipy/spatial/tests/test_distance.py | TestPdist::test_pdist_correlation_iris_float32
9.84s | call | scipy/stats/tests/test_continuous.py | TestTransforms::test_loc_scale

</body></html>

</details>

macOS run passed in 58mins. Here are the slow tests:

<details>

Duration | Module | Test
-- | -- | --
47.58s | scipy/_lib/tests/test_import_cycles.py | test_public_modules_importable_2
35.79s | scipy/special/tests/test_mpmath.py | TestSystematic::test_e1_complex
32.06s | scipy/_lib/tests/test_warnings.py | test_warning_calls_filters (setup)
25.88s | scipy/linalg/tests/test_extending.py | test_cython
23.41s | scipy/special/tests/test_extending.py | test_cython
20.04s | scipy/optimize/tests/test_extending.py | test_cython
19.17s | scipy/special/tests/test_cdflib.py | TestCDFlib::test_chndtrix
18.46s | scipy/special/tests/test_mpmath.py | TestSystematic::test_legendre_int
16.69s | scipy/special/tests/test_mpmath.py | TestSystematic::test_bessely_complex
16.25s | scipy/special/tests/test_mpmath.py | TestSystematic::test_erfc_complex
15.55s | scipy/stats/tests/test_stats.py | TestPageTrendTest::test_accuracy2[542-0.9481266260876332-True-exact-data0]
15.38s | scipy/stats/tests/test_mgc.py | TestMGCStat::test_pvalue_literature
14.73s | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[Logistic]
14.44s | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[_LogUniform]
14.28s | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[StandardNormal]
12.95s | scipy/special/tests/test_mpmath.py | TestSystematic::test_erf_complex
12.69s | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[Normal]
12.54s | scipy/special/tests/test_mpmath.py | TestSystematic::test_pcfw
12.24s | scipy/special/tests/test_hyp2f1.py | TestHyp2f1::test_test_hyp2f1
12.15s | scipy/special/tests/test_mpmath.py | TestSystematic::test_erfi_complex
11.77s | scipy/special/tests/test_mpmath.py | TestSystematic::test_besselk_int
11.43s | scipy/special/tests/test_cdflib.py | TestCDFlib::test_bdtrin
11.35s | scipy/stats/tests/test_multivariate.py | TestNormalInverseGamma::test_moments
10.14s | scipy/special/tests/test_mpmath.py | TestSystematic::test_erfi
9.77s | scipy/special/tests/test_mpmath.py | TestSystematic::test_rgamma
9.50s | scipy/special/tests/test_mpmath.py | TestSystematic::test_log_ndtr_complex
9.21s | scipy/interpolate/tests/test_fitpack2.py | TestRectBivariateSpline::test_spline_large_2d[spl_apis1-s_tols1-shape1]
9.12s | scipy/special/tests/test_mpmath.py | test_beta
8.95s | scipy/stats/tests/test_contingency.py | TestChi2Contingency::test_resampling_randomized[PermutationMethod]
8.82s | scipy/special/tests/test_mpmath.py | test_rgamma_zeros
8.46s | scipy/_lib/tests/test__util.py | TestTransitionToRNG::test_rng_deterministic[basinhopping-seed]
7.93s | scipy/special/tests/test_mpmath.py | test_erf_complex
7.91s | scipy/special/tests/test_cdflib.py | TestCDFlib::test_gdtrib
7.75s | scipy/special/tests/test_cython_special.py | test_cython_api[elliprj]
7.74s | scipy/interpolate/tests/test_fitpack2.py | TestRectBivariateSpline::test_spline_large_2d[spl_apis1-s_tols2-shape1]
7.51s | scipy/special/tests/test_mpmath.py | test_spence_circle
7.46s | scipy/special/tests/test_mpmath.py | TestSystematic::test_expint
7.21s | scipy/special/tests/test_mpmath.py | TestSystematic::test_chi_complex
7.19s | scipy/optimize/tests/test_least_squares.py | TestTRF::test_workers
7.09s | scipy/stats/tests/test_multivariate.py | TestMultivariateNormal::test_cdf_vs_cubature[3]
7.07s | scipy/_lib/tests/test__util.py | TestTransitionToRNG::test_rng_deterministic[differential_evolution-seed]
6.91s | scipy/stats/tests/test_continuous.py | TestTransforms::test_loc_scale
6.89s | scipy/integrate/tests/test_quadrature.py | TestCumulativeSimpson::test_cumulative_simpson_against_simpson_with_default_dx[numpy]
6.87s | scipy/optimize/tests/test_least_squares.py | TestDogbox::test_workers
6.73s | scipy/special/tests/test_mpmath.py | TestSystematic::test_gamma_complex
6.72s | scipy/optimize/tests/test__differential_evolution.py | TestDifferentialEvolutionSolver::test_L1
6.71s | scipy/stats/tests/test_continuous.py | TestDistributions::test_support_moments_sample[Uniform]
6.48s | scipy/optimize/tests/test_milp.py | test_milp_timeout_16545[options1-Iteration limit reached. (HiGHS Status 14:]
6.46s | scipy/special/tests/test_mpmath.py | TestSystematic::test_expm1_complex
6.42s | scipy/optimize/tests/test_constraint_conversion.py | TestNewToOld::test_individual_constraint_objects

</body></html>

</details>

---

I've published `gmpy2` binaries against cp316 to https://prefix.dev/channels/sp-nightly-forge (built on host on my osx-arm64 and linux-64 boxes, and in a `linux/arm64` container via orbstack). Would be very handy to have cross-comp eventually, but that's for another day... @rgommers could you see if `pixi run build-asan` works for you on linux-aarch64 now?

lucascolley commented on pixi.toml:
```diff
@@ -851,6 +856,65 @@ depends-on = "build-system-libs"
+blas-devel = { version = "*", build = "*openblas" }
+
+[feature.asan-build-deps.target.linux-64.dependencies]
+clangxx_linux-64 = "*"
```
ah, probably this needs a companion for `linux-aarch64`

serge-sans-paille commented:
https://github.com/serge-sans-paille/pythran/actions/runs/29151748805 << Minimal Asan checks validating for pythran on x86_64

rgommers commented:
On the last push, macOS took 54m and Linux 1h 24m. Not sure how much further we can push it, might require a faster runner (which also comes with better caching) - I'm inclined to try that in a separate repo.

The other thing that came to mind is sharing the test suite + gating the runs on detecting changes to C/C++/Cython code only. That will be a bit of a project, but diff-dependent testing has come up before and has large benefits - we're going to need it at some point.

rgommers commented:
I'm traveling right now, will try the Linux aarch64 build when I get a chance, hopefully tonight.

serge-sans-paille commented:
Pythran part merged in https://github.com/serge-sans-paille/pythran/pull/2444, nothing detected.

rgommers commented on pixi.toml:
```diff
@@ -851,6 +856,65 @@ depends-on = "build-system-libs"
+
+[feature.asan-build-deps.target.linux-64.dependencies]
+clangxx_linux-64 = "*"
+
```
These two lines are missing:
```
[feature.asan-build-deps.target.linux-aarch64.dependencies]
clangxx_linux-aarch64 = "*
```

After adding them, the build works on Linux aarch64 as well.

rgommers commented:
Test durations on a Linux aarch64 container on an M5 Macbook:
- `pixi r test-asan -j2`: 17m
- `pixi r test-asan -j4 --dist worksteal`: 9m 18s
- `pixi r test-asan -j4 --dist loadfile`: 9m 43s
- `pixi r test-asan -j4` with `malloc_context_size=7`: 9m 34s
- `pixi r test-asan -j8`: 6m

So `pytest-xdist` scaling isn't terrible anymore, but also still not linear.

lucascolley commented on pixi.toml:
```diff
@@ -851,6 +856,65 @@ depends-on = "build-system-libs"
+description = "Test SciPy with address sanitizer"
+
+[feature.asan.target.osx-arm64.tasks.test-asan]
+cmd = "echo $ASAN_OPTIONS && spin test --no-build --build-dir=build-asan --no-capture -m 'not fail_asan'"
```
```suggestion
cmd = "spin test --no-build --build-dir=build-asan --no-capture -m 'not fail_asan'"
```

lucascolley commented on pixi.toml:
```diff
@@ -851,6 +856,68 @@ depends-on = "build-system-libs"
+numpy.subdirectory = "pixi-packages/asan"
+numpy.rev = "6bc853b2fbb3721332de7e5dc44ea9b282bc260a"
+
+gmpy2 = { version = "*", channel = "https://prefix.dev/sp-nightly-forge" }
```
this may be a blocker for merge, not sure we should really be pulling in random binaries that were built on my machine

rgommers replied:
Yeah, it'd be best to avoid that indeed. Is the test deps reorganization to drop `gmpy2` completely here really that bad?

lucascolley replied:
we could re-organise to mirror the `dependency-groups` in `pyproject.toml`, however Pixi features do not have the equivalent of `include-group` which will mean the environment definitions are more verbose.

lucascolley replied:
In case we do go down this road, https://github.com/prefix-dev/pixi/pull/6662 has a fix that allows the following to solve:

```toml
[feature.asan.dependencies]
gmpy2 = {
  git = "https://github.com/gmpy2/gmpy2",
  package = {
    version = "0",
    build = {
      backend.name = "pixi-build-python",
      config = {
        compilers = ["c"],
        ignore-pypi-mapping = false,
      }
    },
    host-dependencies = {
      gmp = "*",
      mpfr = "*",
      mpc = "*",
      python = {
        flags = ["asan"],
        git = "https://github.com/lucascolley/cpython",
        subdirectory = "Tools/pixi-packages",
        rev = "8b5b0c29797cf88d78ef014916a5e5a5d51bbf95",
      },
    },
  },
}
```

lucascolley commented on pixi.toml:
```diff
@@ -851,6 +856,68 @@ depends-on = "build-system-libs"
+numpy.flags = ["asan"]
+numpy.git = "https://github.com/numpy/numpy"
+numpy.subdirectory = "pixi-packages/asan"
+numpy.rev = "6bc853b2fbb3721332de7e5dc44ea9b282bc260a"
```
once we can pin to revs on the main branch of CPython and NumPy, we should be able to merge

lucascolley commented on pixi.toml:
```diff
@@ -851,6 +856,68 @@ depends-on = "build-system-libs"
+clangxx_linux-aarch64 = "*"
+
+[feature.asan-build-deps.target.osx-arm64.dependencies]
+clangxx_osx-arm64 = "*"
```
side comment: would be nice to have a single meta package in conda-forge for this

EDIT: asked at [#compilers > activated &#96;clangxx&#96; metapackage?](https://conda-forge.zulipchat.com/#narrow/channel/516932-compilers/topic/activated.20.60clangxx.60.20metapackage.3F/with/611029328)

lucascolley replied:
I suppose this is an improvement:

```diff
diff --git a/pixi.toml b/pixi.toml
index 5d2ffd42b3..39d2ec9f1a 100644
--- a/pixi.toml
+++ b/pixi.toml
@@ -888,19 +888,15 @@ pybind11 = "*"
 numpy = "*"
 blas-devel = { version = "*", build = "*openblas" }

-[feature.asan-build-deps.target.linux-64.dependencies]
-clangxx_linux-64 = "*"
-
-[feature.asan-build-deps.target.linux-aarch64.dependencies]
-clangxx_linux-aarch64 = "*"
-
-[feature.asan-build-deps.target.osx-arm64.dependencies]
-clangxx_osx-arm64 = "*"
+[feature.asan-build-deps.dependencies]
+clangxx = "*"

 [feature.asan.tasks.build-asan]
 # `use-pythran=false` needed with py316dev due to https://github.com/serge-sans-paille/gast/issues/110
 cmd = "spin build --asan --build-dir=build-asan --setup-args=-Duse-pythran=false --setup-args=-Dblas=blas --setup-args=-Dlapack=lapack --setup-args=-Duse-g77-abi=true"
 env.ASAN_OPTIONS = "detect_leaks=0:strict_init_order=true:use_sigaltstack=0"
+env.CC = "clang"
+env.CXX = "clang++"
 description = "Build SciPy, NumPy, and CPython with address sanitizer"
```

rgommers commented:
The two `tmfqr` failures should go away when gh-25632 is merged.

rgommers commented:
I've so far hit an OpenBLAS bug in `sdot` (already fixed, linked above) and now what looks like a bug in the AVX512 `log`/`exp` code in NumPy. The latter in CI on Blacksmith that's reasonably fast: https://github.com/scipy-ci/scipy/actions

It's going to take a little while to work through the problems here. I think we should probably get this merged _without_ a CI job @lucascolley, since that job will be too slow on standard GitHub runners. In the `scipy-ci/scipy` repo we can experiment more easily, it'd be good to get it green reliably with jobs in the <=20 min range.

rgommers commented on pixi.toml:
```diff
@@ -851,6 +856,68 @@ depends-on = "build-system-libs"
+description = "Test SciPy with address sanitizer"
+
+[feature.asan.target.osx-arm64.tasks.test-asan]
+cmd = "spin test --no-build --build-dir=build-asan --no-capture -m 'not fail_asan'"
```
`-m 'not fail_asan'` overrides the default test selection, so it ends up running the full rather than the fast test suite. We need the latter. Can you make it `not slow and not fail_asan` and see how long it takes then? 

For reference, the test suite took about 18 minutes with `-j2` on a Blacksmith runner.

lucascolley commented:
Hmm, `bin/numpy-config` is present in the built NumPy package but not found by meson on linux-64

lucascolley commented on pixi.toml:
```diff
@@ -850,6 +855,68 @@ depends-on = "build-system-libs"
+
+[feature.asan.tasks.build-asan]
+# `use-pythran=false` needed with py316dev due to https://github.com/serge-sans-paille/gast/issues/110
+cmd = "ls $CONDA_PREFIX/bin && echo $PATH && spin build --asan --build-dir=build-asan --setup-args=-Duse-pythran=false --setup-args=-Dblas=blas --setup-args=-Dlapack=lapack --setup-args=-Duse-g77-abi=true"
```
this is bizarre — CI now seems to confirm that `numpy-config` is on `PATH`, but meson still doesn't find it.

rgommers replied:
Executable bit getting lost somehow? The `x` at the end of the permissions:
```
ls -l .pixi/envs/build/bin/*numpy*
-rwxr-xr-x@ 1 rgommers  staff  286 Nov 17  2025 .pixi/envs/build/bin/numpy-config
```

lucascolley replied:
> Executable bit getting lost somehow?

that's not it:

```console
scipy on 🎋 asan-stack [⇕] is 📦 v1.19.0.dev0 via 🐍 v3.16.0a0 via 🧚 v0.73.0 (asan)
❯ ls /home/lucas/ghq/github.com/scipy/scipy/.pixi/envs/asan/bin/numpy-config -l | get name mode
╭───┬─────────────────────────────────────────────────────────────────────────────────╮
│ 0 │ ╭───┬─────────────────────────────────────────────────────────────────────────╮ │
│   │ │ 0 │ /home/lucas/ghq/github.com/scipy/scipy/.pixi/envs/asan/bin/numpy-config │ │
│   │ ╰───┴─────────────────────────────────────────────────────────────────────────╯ │
│ 1 │ ╭───┬───────────╮                                                               │
│   │ │ 0 │ rwxr-xr-x │                                                               │
│   │ ╰───┴───────────╯                                                               │
╰───┴─────────────────────────────────────────────────────────────────────────────────╯

scipy on 🎋 asan-stack [🔨⇕] is 📦 v1.19.0.dev0 via 🐍 v3.16.0a0 via 🧚 v0.73.0 (asan)
❯ which numpy-config
╭───┬──────────────┬─────────────────────────────────────────────────────────────────────────┬──────────╮
│ # │   command    │                                  path                                   │   type   │
├───┼──────────────┼─────────────────────────────────────────────────────────────────────────┼──────────┤
│ 0 │ numpy-config │ /home/lucas/ghq/github.com/scipy/scipy/.pixi/envs/asan/bin/numpy-config │ external │
╰───┴──────────────┴─────────────────────────────────────────────────────────────────────────┴──────────╯
```

lucascolley replied:
ah, but that is instructive:

```
❯ numpy-config
==112348==Your application is linked against incompatible ASan runtimes.
```

---

```
│ │ [lib/python3.16/site-packages/numpy/random/_philox.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  └─ libc.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/mtrand.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libm.so.6 (system)
 │ │  ├─ libasan.so.8
 │ │  └─ libc.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/linalg/_umath_linalg.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libopenblas.so.0
 │ │  ├─ libc.so.6 (system)
 │ │  ├─ libm.so.6 (system)
 │ │  └─ libasan.so.8
 │ │
 │ │ [lib/python3.16/site-packages/numpy/_core/_multiarray_umath.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  ├─ libc.so.6 (system)
 │ │  ├─ lib/libstdc++.so.6.0.34 (libstdcxx)
 │ │  ├─ lib/libgcc_s.so.1 (libgcc)
 │ │  ├─ libm.so.6 (system)
 │ │  └─ libopenblas.so.0
 │ │
 │ │ [lib/python3.16/site-packages/numpy/_core/_struct_ufunc_tests.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  └─ libc.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/_core/_rational_tests.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  └─ libc.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/fft/_pocketfft_umath.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ lib/libstdc++.so.6.0.34 (libstdcxx)
 │ │  ├─ libasan.so.8
 │ │  ├─ libc.so.6 (system)
 │ │  └─ lib/libgcc_s.so.1 (libgcc)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/_pcg64.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libasan.so.8
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/bit_generator.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libasan.so.8
 │ │
 │ │ [lib/python3.16/site-packages/numpy/_core/_multiarray_tests.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libm.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/_sfc64.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  └─ libc.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/_common.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  └─ libc.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/_mt19937.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libasan.so.8
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/_bounded_integers.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libm.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/linalg/lapack_lite.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libopenblas.so.0
 │ │
 │ │ [lib/python3.16/site-packages/numpy/_core/_umath_tests.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libc.so.6 (system)
 │ │  ├─ libm.so.6 (system)
 │ │  └─ libasan.so.8
 │ │
 │ │ [lib/python3.16/site-packages/numpy/_core/_simd.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libasan.so.8
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libm.so.6 (system)
 │ │
 │ │ [lib/python3.16/site-packages/numpy/_core/_operand_flag_tests.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  └─ libasan.so.8
 │ │
 │ │ [lib/python3.16/site-packages/numpy/random/_generator.cpython-316-x86_64-linux-gnu.so] links against:
 │ │  ├─ libm.so.6 (system)
 │ │  ├─ libc.so.6 (system)
 │ │  └─ libasan.so.8
```

lucascolley replied:
fixed this by introducing `workspace.build-variants` in this repo. @rgommers contrary to our discussion at EuroPython, it actually is the case that build variants are determined by the workspace of the environment and propagated through for source dependencies, so for our use case the `workspace.build-variants` in the NumPy and CPython workspaces are superfluous.

Still, I think it is helpful to track them in each repo, to show what we intend to support and what ought to be usable downstream.

---

Sidenote: I had to delete the lock file on this branch and re-lock from the file on main to unblock this and get the correct variants to install.

rgommers replied:
> it actually is the case that build variants are determined by the workspace of the environment and propagated through for source dependencies

Interesting. I looked through the pixi docs, but the various mentions of `[orkspace.build-variants]` don't quite explain this well AFAICT. Is it an arbitrary override thing, or are `c-compiler`/`cxx-compiler` entries special?

lucascolley replied:
https://github.com/prefix-dev/pixi/issues/6677

lucascolley commented on pixi.toml:
```diff
@@ -24,6 +24,11 @@ version = "*"
 [package.build.config]
 compilers = ["c", "cxx"]
 
+[workspace.build-variants]
+# NOTE: this is necessary for now to get all source packages in the ASAN env to use clang
+c_compiler = ["clang"]
+cxx_compiler = ["clangxx"]
```
ref. https://github.com/scipy/scipy/pull/24066#discussion_r3646805316

lucascolley commented:
The ubuntu job is catching some problem in `fast_matrix_market`:
```
AddressSanitizer: CHECK failed: asan_interceptors.cpp:463 "((__interception::real___cxa_throw)) != (0)" (0x0, 0x0) (tid=21883)
    #0 0x55dc8c2b7c67  (/home/runner/work/scipy/scipy/.pixi/envs/asan/bin/python3.16+0x12ac67)
    #1 0x55dc8c2cf6ac  (/home/runner/work/scipy/scipy/.pixi/envs/asan/bin/python3.16+0x1426ac)
    #2 0x55dc8c2a3d9d  (/home/runner/work/scipy/scipy/.pixi/envs/asan/bin/python3.16+0x116d9d)
    #3 0x7b8f50c59b54  (/home/runner/work/scipy/scipy/build-asan-install/usr/lib/python3.16/site-packages/scipy/io/_fast_matrix_market/_fmm_core.cpython-316-x86_64-linux-gnu.so+0x59b54)
    #4 0x7b8f50c4e96c  (/home/runner/work/scipy/scipy/build-asan-install/usr/lib/python3.16/site-packages/scipy/io/_fast_matrix_market/_fmm_core.cpython-316-x86_64-linux-gnu.so+0x4e96c)
    #5 0x7b8f50cc847c  (/home/runner/work/scipy/scipy/build-asan-install/usr/lib/python3.16/site-packages/scipy/io/_fast_matrix_market/_fmm_core.cpython-316-x86_64-linux-gnu.so+0xc847c)
    #6 0x7b8f50cc8293  (/home/runner/work/scipy/scipy/build-asan-install/usr/lib/python3.16/site-packages/scipy/io/_fast_matrix_market/_fmm_core.cpython-316-x86_64-linux-gnu.so+0xc8293)
    #7 0x7b8f50c723c6  (/home/runner/work/scipy/scipy/build-asan-install/usr/lib/python3.16/site-packages/scipy/io/_fast_matrix_market/_fmm_core.cpython-316-x86_64-linux-gnu.so+0x723c6)
```

lucascolley commented on pixi.toml:
```diff
@@ -850,6 +862,91 @@ depends-on = "build-system-libs"
+# point asan's `__cxa_throw` interceptor has already been initialized with a null "real"
+# pointer, so any C++ exception thrown from an extension aborts with
+# "CHECK failed: asan_interceptors.cpp:463 ... real___cxa_throw ... != (0)"
+env.LD_PRELOAD = "$PIXI_PROJECT_ROOT/.pixi/envs/asan/lib/libstdc++.so.6"
```
I believe this wasn't always needed during this PR's development...

lucascolley commented:
@rgommers I fixed the linux-64 job, it just passed in 38mins

lucascolley commented:
With setup-pixi caching online, these jobs are now not the slowest, and that is without even having ccache set up.

rgommers commented:
@lucascolley how about dropping all CI changes here, merging this PR, then reopening a separate PR with CI changes. Those are the most critical and blocking right now (our CI load is really high right now, and this piles on more - also there are still TODOs about caching in the PR). And they're the least important - it'd be really helpful to get a `test-asan` task available locally.

lucascolley committed:
BLD/DEV: ASAN at your fingertips

[skip ci]

lucascolley commented on pixi.toml:
```diff
@@ -854,6 +866,95 @@ depends-on = "build-system-libs"
+  # point asan's `__cxa_throw` interceptor has already been initialized with a null "real"
+  # pointer, so any C++ exception thrown from an extension aborts with
+  # "CHECK failed: asan_interceptors.cpp:463 ... real___cxa_throw ... != (0)"
+  LD_PRELOAD = "$PIXI_PROJECT_ROOT/.pixi/envs/asan/lib/libstdc++.so.6",
```
I'm still not sure why this is needed now but didn't seem to be needed before — the fix is Claude's.

isuruf replied:
Can you run with `LD_DEBUG=libs` without `LD_PRELOAD` and give the output?

lucascolley replied:
Here's a gist with the full output up to the crash that this workaround fixes: https://gist.github.com/lucascolley/d6b82cf60a77086e7b0d0d99699b44f3

isuruf replied:
Ah, you should link to shared libasan with `-shared-libsan`.

lucascolley replied:
> Ah, you should link to shared libasan with `-shared-libsan`.

which repo is that a fix to?

lucascolley replied:
> Ah, you should link to shared libasan with `-shared-libsan`.

For the SciPy build, we specify `-Db_sanitize=address` and Meson is responsible for adding `-fsanitize=address`. Do you mean that Meson should be adding `-shared-libsan` too?

isuruf replied:
You can add it in `python` recipe and I assume meson will propagate the cflags from python to extensions.

rgommers replied:
> I assume meson will propagate the cflags from python to extensions.

This won't happen. No build backends except for `setuptools` propagate the CFLAGS that `python` was built with (for good reasons, that's inherently kinda broken).

If you want a compile flag picked up, you should put it in the relevant CFLAGS/CXXFLAGS/LDFLAGS.

lucascolley replied:
I opened https://github.com/python/cpython/issues/155375. Do I understand correctly, then, that we would need both that merged, and adding it to `LDFLAGS` in the SciPy manifest here?

rgommers replied:
Yes indeed - and NumPy too then I'd think.

lucascolley reviewed (commented):
@rgommers yes, sounds good! I'd appreciate your input on the three 'outstanding comments' in the top-post before merge.

lucascolley commented:
@rgommers please could you give `pixi run test-asan` another go with the current state of this branch? All seems to be working for me on native linux-64, native osx-arm64, and a linux-arm64 container. If you agree, then I think we should push for the remaining fixes to be upstreamed to CPython and NumPy, then we can finally merge this
